pkg/pool-stable-phantom/contracts/StablePhantomPool.sol

```diff
@@ -578,9 +578,60 @@ contract StablePhantomPool is IRateProvider, BaseGeneralPool, ProtocolFeeCache {
 
         if (kind == StablePhantomPoolUserData.JoinKindPhantom.COLLECT_PROTOCOL_FEES) {
             return _collectProtocolFees();
+        } else if (kind == StablePhantomPoolUserData.JoinKindPhantom.EXACT_TOKENS_IN_FOR_BPT_OUT) {
+            return _joinExactTokensInForBPTOut(balances, scalingFactors, protocolSwapFeePercentage, userData);
+        } else {
+            _revert(Errors.UNHANDLED_JOIN_KIND);
         }
+    }
 
-        _revert(Errors.UNHANDLED_BY_PHANTOM_POOL);
+    function _joinExactTokensInForBPTOut(
+        uint256[] memory balances,
+        uint256[] memory scalingFactors,
+        uint256 protocolSwapFeePercentage,
+        bytes memory userData
+    )
+        private
+        returns (
+            uint256,
+            uint256[] memory,
+            uint256[] memory
+        )
+    {
+        (uint256[] memory amountsIn, uint256 minBPTAmountOut) = userData.exactTokensInForBptOut();
+        InputHelpers.ensureInputLengthMatch(_getTotalTokens() - 1, amountsIn.length);
+
+        uint256[] memory scaledAmountsInWithBpt = _addBptItem(amountsIn, 0);
+        _upscaleArray(scaledAmountsInWithBpt, scalingFactors);
+
+        // Balances come in from onJoinPool already upscaled
+        uint256 bptAmountOut = _computeBptAmountOut(balances, scaledAmountsInWithBpt);
+        _require(bptAmountOut >= minBPTAmountOut, Errors.BPT_OUT_MIN_AMOUNT);
+
+        if (protocolSwapFeePercentage > 0) {
+            _trackDueProtocolFeeByBpt(bptAmountOut, protocolSwapFeePercentage);
+        }
+
+        return (bptAmountOut, _addBptItem(amountsIn, 0), new uint256[](balances.length));
```

Comment:
```suggestion
        return (bptAmountOut, scaledAmountsInWithBpt, new uint256[](balances.length));
```
We're currently returning unscaled amounts back into `_onJoinPool` (which expects scaled values to be returned).